### DISCUSSION TITLE
Modularisierung der Datapoint-Konfiguration durch Migration in eine relationale SQLite-Datenbank

### DISCUSSION THREAD

**Discussion by hillbicks** [open3e/open3e#129] [Ideas]
Hi zusammen,

ich habe vor einiger Zeit meine Vitocal 252 in Betrieb genommen, die Anbindung erfolgt dabei an Home Assistant. Habe mich dazu des repos von Wolfang bedient, der seine 250 so manuell in Home Assistant eingebunden hat. In diesem Zuge habe ich mir Gedanken darueber gemacht, wie der Aufbau der DIDs und der gesamten Config vielleicht etwas einfacher zu handhaben sein koennte. Da die config mittlerweile schon ueber 34.000 Zeilen hat, schlage ich die Speicherung der Configuration in einer relationalen sqlite Datenbank vor 

![output](https://github.com/user-attachments/assets/87ece37f-add3-4f09-8e58-052f27544b5f)

Das soll nur ein Beispiel sein um zu verdeutlichen wie ich mir das vorstelle. Home Assistant bietet eine auto discovery um mqtt devices automatisch anzulegen. Dazu muessen die Daten im json Format an das entsprechende Topic gesendet werden. Man koennte die Datenbank daher so aufbauen, dass das json automatisch aus den vorhanden Daten zusammen gebaut werden kann. 

Sofern hier bei den Entwicklern ein Interesse daran besteht, dann kann ich mich da gerne noch weiter mit beschaeftigen und konkretere Vorschlaege machen. 

Insgesamt wuerde das die Wartbarkeit und Anbindungsmoeglichkeiten wesentlich erleichtern. Im naechsten Schritt koennte man dann noch ein Template fuer fehlende DIDs erstellen die von den Usern ausgefuellt werden muessen um sie dann halb automatisiert in die DB zu schreiben. 

Freue mich ueber euer Feedback

**Comment by philippoo66** [comment]
moin @hillbicks !

ich hab das grad erst gesehen... ich nehme an, das bezieht sich im Kern auf Home Assistant? da kann ich nix zu sagen (kann ich ja eh eigentlich grundsätzlich nicht, weil ich garkein E3 device habe), weil ich keine Ahnung von Home Assistant habe.

Für open3e ist es definitiv keine Option, weil SQL ein recht komplexes Gebilde ist, und ich das Funktionieren von open3e nicht ansowas gebunden sehe. Das ist so schon alles kompliziert genug...

Der 'Normalbenutzer' ist ja erstmal auch nicht mit den Datapoints Dateien konfrontiert, und einem Rechner ist es ziemlich egal, wie gross die ist, wobei ich da (open3e) auch nur knapp 1.700 Zeilen sehe und keine 34.000...

beste Grüsse!
Phil

**Comment by MyHomeMyData** [comment]
@hillbicks @philippoo66 Ich finde, wir sollten den Vorschlag diskutieren. Wir wollen ja die Integration von open3e in HA auf jeden Fall weiter verbessern. Allerdings ist - zumindest mir - noch nicht klar, was das beste Konzept dafür ist.

Ist es überhaupt notwendig, die SQLite-DB in open3e zu benutzen? Oder reicht es, passend strukturierte MQTT-Messages zu verarbeiten? Bzgl. Komplexität bin ich auf jeden Fall bei Dir, Phil. Wir sollten versuchen, open3e schlank zu halten.

@jwende @Philip-Wiege Wie seht ihr das?

**Reply by Philip-Wiege** [reply]
Hi zusammen,
ich bin hier tatsächlich auch etwas überfordert. Soll hier quasi das gesamte interne Datenmodell von Open3e umgebaut werden?

Die Frage ist, ob Open3e eine eierlegende Wollmilchsau werden soll, oder eher ein solider Kern für alle Nebenprojekte.

Ich sehe es mittlerweile so, dass Open3e doch als Python Package auch von anderer Python SW verwendet werden kann oder? Ich würde daher meine ersten Versuche einer "Brücke" zu Home Assistant als eigenes Python Projekt sehen, welches die open3e "library" verwendet.

Momentan bietet open3e ja über das schreiben und lesen hinaus den Mehrwert der Command Queue und des MQTT Client.

**Comment by hillbicks** [comment]
Danke erstmal fuers Feedback, ich versuche grade noch was Klarheit in den Vorschlag zu bringen. 

HomeAssistant habe ich nur als Beispiel genommen, weil das mein System der Wahl ist. Aus Sicht des Endusers ergibt sich eigentlich immer das gleiche Szenario: Ich habe Waermepumpe XYZ und moechte a) ausgewaehlte/alle Daten der Waermepumpe in der Software meiner Wahl haben und b) Daten veraendern/Befehle senden koennen (ebenfalls ueber besagte Software)

Was mir  vorschwebt ist ne Single Source of Truth aus der sich u.a. diese Infos ableiten lassen:
- Datentyp der einzelnen DIDs
- Welche DIDs gehoeren zu Device XYZ
- Anreichern der DIDs mit weiteren Datentypen, z.b. type: temperature oder humidity.

Letzendlich eine Datenbasis schaffen mit der dann die umliegenden Systeme arbeiten koennen. In diesem Fall koennte man also aus dieser Datenbasis eine MQTT message zusammenbauen die dann home assistant verarbeiten kann. iobroker arbeitet vermutlich anders um neue Devices anzulegen, wuerde aber auf die gleiche Datenbasis zugreifen. Bestandsdatenpflege als auch Neuanlage waeren somit einfach an 3rd party system zu uebertragen. 

PS: Mein Hirn steht grade kurz vor Tilt, ich stecke kurz vor Ende der Sanierung und Start vom Umzug, seht mir das wirre Gebrabbel bitte nach :p

**Reply by philippoo66** [reply]
> Datentyp der einzelnen DIDs

genau das ist ja in den codecs aufgeschlüsselt.

> Welche DIDs gehoeren zu Device XYZ

das findet depict_system heraus und das kannst du in keiner Datenbank zuordnen weil es sich auch schon mal mit einem Firmware uptdate ändert und nicht alle die gleiche Firmware fahren. Ausserdem wäre es von Nachteil, wenn man jedes mal irgendwas an open3e machen müsste, wenn Viessmann mal wiedr ne Firmware rausbringt. Aus diesem Grund sind wir ja weg von dem -dev Konzept... Wir kennen auch nicht alle Geräte.

> Anreichern der DIDs mit weiteren Datentypen, z.b. type: temperature oder humidity.

das ist jetzt wirklich ziemlich wirr. Temperatur und Feuchte sind keine Datentypen. Du kannst einen DID auch nicht verändern, der kommt ja so wie Viessmann ihn 'gemacht' hat.

Kann es sein, dass du dich erstmal noch etwas eingehender mit der Geschichte beschäftigen musst?! Wenn dir was nicht klar ist, helfen wir gern :-)

Ist dir eigentlich bewusst, das udsoncan die Codecs genau so haben will, wie sie in den Datapoints_... .py drin stehen? Dass also jedes andere 'Format' erstmal wieder darein 'übersetzt' werden müsste?!

**Reply by philippoo66** [reply]
ps. oder verstehe ich dich vielleicht völlig falsch und es geht weniger um open3e als um HA und ich blicke das nicht weil ich da garkeinen Schimmer von habe?
